source/Calamari/Scripts/AzureContext.ps1

```diff
@@ -60,58 +60,109 @@ function Execute-WithRetry([ScriptBlock] $command) {
-
-                # Force any output generated to be verbose in Octopus logs.
-                Write-Host "##octopus[stdout-verbose]"
-                Login-AzureRmAccount -Credential $creds -TenantId $OctopusAzureADTenantId -SubscriptionId $OctopusAzureSubscriptionId -Environment $AzureEnvironment -ServicePrincipal
-                Write-Host "##octopus[stdout-default]"
-            }
-            elseif (Get-InstalledModule Az -ErrorAction SilentlyContinue)
-            {
-                if (-Not(Get-Command "Disable-AzureRMContextAutosave" -errorAction SilentlyContinue))
-                {
-                    # Turn on AzureRm aliasing
-                    # See https://docs.microsoft.com/en-us/powershell/azure/migrate-from-azurerm-to-az?view=azps-3.0.0#enable-azurerm-compatibility-aliases
-                    Enable-AzureRmAlias -Scope Process
+            # Depending on which version of Powershell we are running under will change which module context we want to initialize.            
+            #   Powershell Core: Check Az then AzureRM (provide a warning and do nothing if AzureRM is installed)
+            #   Windows Powershell: Check AzureRM then Az
+            #
+            # If a module is installed (e.g. AzureRM) then testing for it causes the module to be loaded along with its various assemblies
+            # and dependencies. If we then test for the other module (e.g. Az) then it also loads it's module and assemblies which then
+            # creates havoc when you try and call any context methods such as Disable-AzContextAutosave due to version differences etc. 
+            # For this reason we'll only test the module we prefer and then if it exists initialize it and not the other one.
+            if (Get-RunningInPowershellCore) {
+                if (Get-AzModuleInstalled) {
+                    Initialize-AzContext
                 }
-
-                # Turn off context autosave, as this will make all authentication occur in memory, and isolate each session from the context changes in other sessions
-                Disable-AzContextAutosave -Scope Process
-
-                $AzureEnvironment = Get-AzEnvironment -Name $OctopusAzureEnvironment
-                if (!$AzureEnvironment)
-                {
-                    Write-Error "No Azure environment could be matched given the name $OctopusAzureEnvironment"
-                    exit -2
+                elseif (Get-AzureRmModuleInstalled) {
+                    # AzureRM is not supported on powershell core
+                    Write-Warning "AzureRM module is not compatible with Powershell Core, authentication will not be performed with AzureRM"
+                }                
+            }            
+            else { 
+                # Windows Powershell
+                if (Get-AzureRmModuleInstalled) {
+                    Initialize-AzureRmContext
+                }
+                elseif (Get-AzModuleInstalled) {
```

Comment:
Should az tools be the first one, given RM is dead?

Reply:
I left it this way to maintain backward compatibility with the previous behaviour when running under Windows Powershell. There probably needs to be a discussion about what to do with AzureRM though, I'll raise it and see if we want to flip the ordering around here.